lib/SILOptimizer/Mandatory/Differentiation.cpp

```diff
@@ -6660,13 +6661,22 @@ class PullbackEmitter final : public SILInstructionVisitor<PullbackEmitter> {
       // Get predecessor terminator operands.
       SmallVector<std::pair<SILBasicBlock *, SILValue>, 4> incomingValues;
       bbArg->getSingleTerminatorOperands(incomingValues);
-      // Initialize adjoint value of predecessor terminator operands as
-      // adjoint value of current block arguments.
+      // Materialize adjoint value of predecessor terminator operands, create a
+      // copy, and and set copy as adjoint value of current block arguments.
       auto bbArgAdj = getAdjointValue(bb, bbArg);
+      auto concreteBBArgAdj = materializeAdjoint(bbArgAdj, pbLoc);
+      auto concreteBBArgAdjCopy =
+          builder.emitCopyValueOperation(pbLoc, concreteBBArgAdj);
       for (auto pair : incomingValues) {
         auto *predBB = std::get<0>(pair);
         auto incomingValue = std::get<1>(pair);
-        setAdjointValue(predBB, incomingValue, bbArgAdj);
+        // FIXME: This is a hack because `blockTemporarySet` cannot be updated.
+        // Consider changing `blockTemporarySet` to be per-bb like
+        // `blockTemporaries`?
+        blockTemporaries[getPullbackBlock(predBB)].push_back(
```

Comment:
Note: here is the key fix, though it's not robust.

Directly updating `blockTemporaries` without updating `blockTemporarySet` is not robust. However, `blockTemporarySet` currently has type `DenseSet<SILValue>` and tracks the temporary values for the currently visited bb (`bb`). This is too limiting because we want to register a temporary for a different bb (`getPullbackBlock(predBB)`).

---

Tentative fix: change `blockTemporarySet` to be `DenseMap<SILBasicBlock *, llvm::DenseSet<SILValue>>`? This makes it like `blockTemporaries`, which has type `DenseMap<SILBasicBlock *, SmallVector<SILValue, 64>>`.

Or, we can unify `blockTemporaries` and `blockTemporarySet` with a single `DenseMap<SILBasicBlock *, SetVector<SILValue>>`?

cc @rxwei

Reply:
Any issues with simply calling `recordTemporary`?

Reply:
> Any issues with simply calling `recordTemporary`?

`concreteBBArgAdjCopy` is the copied `AdjointValue` to be used in pullback successor bbs.

Calling `recordTemporary(concreteBBArgAdjCopy)` would register `concreteBBArgAdjCopy` as a temporary value in `getPullbackBlock(bb)`, which is not desirable. We want to register `concreteBBArgAdjCopy` as a temporary value in `getPullbackBlock(predBB)` instead.

Reply:
Unified `blockTemporaries` and `blockTemporarySet` as a single DenseMap<SILBasicBlock *, SmallSetVector<SILValue, 64>> in https://github.com/apple/swift/pull/28253/commits/5bb593f084fd1f61c49845264bc4f276b3e8b0aa. I believe this makes the code robust.

Reply:
Ah I see. I would suggest adding a separate method that's tied to this particular use case.

Reply:
I'm not sure why a nested `SetVector` is better than adding a lightweight method. This is slightly inefficient. I don't have a strong preference though.

Reply:
> Ah I see. I would suggest adding a separate method that's tied to this particular use case.

Adding a `recordTemporary(SILBasicBlock *, SILValue)` helper would require changing `blockTemporarySet` from `DenseSet<SILValue>>` to `DenseMap<SILBasicBlock *, DenseSet<SILValue>>`.

Instead, I found unifying `blockTemporaries` and `blockTemporarySet` to be simpler than maintaining two data structures. We can directly use `SetVector` APIs.